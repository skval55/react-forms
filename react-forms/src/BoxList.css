.BoxList-boxContainer {
    margin: 20px;
    display: flex;
    gap: 4px;
    flex-wrap: wrap;
    justify-content: center;
}