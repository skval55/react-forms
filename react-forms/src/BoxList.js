import React, { useState } from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import "./BoxList.css";
import { v4 as uuid } from "uuid";

const Boxlist = () => {
  const [boxes, setBoxes] = useState([]);

  const renderBoxes = () => {
    return (
      <div className="BoxList-boxContainer">
        {boxes.map((box) => (
          <Box
            key={box.code}
            code={box.code}
            height={box.height}
            width={box.width}
            backgroundColor={box.backgroundColor}
            handleDelete={handleDelete}
          />
        ))}
      </div>
    );
  };
  /** Add new Box */
  const addBox = (box) => {
    let newBox = { ...box, code: uuid() };
    setBoxes((boxes) => [...boxes, newBox]);
  };

  function handleDelete(code) {
    setBoxes((boxes) => boxes.filter((box) => box.code !== code));
  }

  return (
    <>
      <h1>Add a Box</h1>
      <NewBoxForm addBox={addBox} />
      {renderBoxes()}
    </>
  );
};

export default Boxlist;
