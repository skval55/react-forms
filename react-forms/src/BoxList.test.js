import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

// smoke test
test("renders without crashing", () => {
  render(<BoxList />);
});

// snapshot test
it("matches snapshot", () => {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("can add a box from the form", function () {
  const { getByLabelText, queryByText } = render(<BoxList />);

  expect(queryByText("X")).not.toBeInTheDocument();

  const heightInput = getByLabelText("height:");
  const widthInput = getByLabelText("width:");
  const backgroundColorInput = getByLabelText("Background Color:");
  const submitBtn = queryByText("Add a new box");

  fireEvent.change(heightInput, { target: { value: 60 } });
  fireEvent.change(widthInput, { target: { value: 100 } });
  fireEvent.change(backgroundColorInput, { target: { value: "blue" } });
  fireEvent.click(submitBtn);

  expect(queryByText("X")).toBeInTheDocument();
});

it("can delete a box", function () {
  const { getByLabelText, queryByText } = render(<BoxList />);

  //   nothing in dom
  expect(queryByText("X")).not.toBeInTheDocument();

  const heightInput = getByLabelText("height:");
  const widthInput = getByLabelText("width:");
  const backgroundColorInput = getByLabelText("Background Color:");
  const submitBtn = queryByText("Add a new box");

  //   add box
  fireEvent.change(heightInput, { target: { value: 60 } });
  fireEvent.change(widthInput, { target: { value: 100 } });
  fireEvent.change(backgroundColorInput, { target: { value: "blue" } });
  fireEvent.click(submitBtn);

  expect(queryByText("X")).toBeInTheDocument();

  //   click del button
  fireEvent.click(queryByText("X"));

  //   again nothing in dom
  expect(queryByText("X")).not.toBeInTheDocument();
});
