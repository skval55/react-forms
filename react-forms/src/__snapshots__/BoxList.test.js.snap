// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <h1>
    Add a Box
  </h1>
  <form
    class="BoxForm"
  >
    <label
      class="BoxForm-label"
      for="height"
    >
      height:
    </label>
    <input
      class="BoxForm-input"
      id="height"
      name="height"
      value=""
    />
    <label
      class="BoxForm-label"
      for="width"
    >
      width:
    </label>
    <input
      class="BoxForm-input"
      id="width"
      name="width"
      value=""
    />
    <label
      class="BoxForm-label"
      for="backgroundColor"
    >
      Background Color:
    </label>
    <input
      class="BoxForm-input"
      id="backgroundColor"
      name="backgroundColor"
      value=""
    />
    <button
      class="BoxForm-button"
    >
      Add a new box
    </button>
  </form>
  <div
    class="BoxList-boxContainer"
  />
</DocumentFragment>
`;
